README
BATMAN-ADV14-LEGACY
-----------------

This is the batman-adv14 "legacy" version (compat-level 14). Use it
if you have a mesh using compat 14 which can't be easily upgraded
to  a newer version of batman-adv14. If you are building a new mesh
network, you should consider the current upstream version,  which
can be found under

http://www.open-mesh.org/projects/batman-adv14/wiki

Please  don't  bother the upstream batman-adv14 developers with bug
reports if you are using batman-adv14-legacy. The batman-adv14-legacy
maintainers can be found at the bottom of this README.

BRANCHES
--------

There are currently two branches in this repository:

* master
* legacy-vis

The in-kernel mesh visualization support ("vis_mode" option) has
been  removed  from  the master branch, so a userspace tool like
alfred has to be used. If you still require "vis_mode", you  can
use the legacy-vis branch.

These branches might be merged in the future to allow "vis_mode"
to be enabled or disabled at runtime.

BATMAN-ADV14
----------

Batman  advanced  is  a new approach to wireless networking which
does no longer operate on the IP basis. Unlike the batman daemon,
which  exchanges  information  using UDP packets and sets routing
tables, batman-adv14anced operates on ISO/OSI Layer 2 only and uses
and  routes  (or  better: bridges) Ethernet Frames. It emulates a
virtual network switch of all nodes participating.  Therefore all
nodes  appear  to be link local, thus all higher operating proto-
cols won't be affected by any changes within the network. You can
run almost any protocol above batman advanced, prominent examples
are: IPv4, IPv6, DHCP, IPX.

Batman advanced was implemented as a Linux kernel driver  to  re-
duce the overhead to a minimum. It does not depend on any (other)
network driver, and can be used on wifi as well as ethernet  lan,
vpn,  etc ... (anything with ethernet-style layer 2).


CONFIGURATION
-------------

Load the batman-adv14 module into your kernel:

# insmod batman-adv14.ko

The  module  is now waiting for activation. You must add some in-
terfaces on which batman can operate. After  loading  the  module
batman  advanced  will scan your systems interfaces to search for
compatible interfaces. Once found, it will create  subfolders  in
the /sys directories of each supported interface, e.g.

# ls /sys/class/net/eth0/batman_adv14/
# iface_status  mesh_iface

If an interface does not have the "batman_adv14" subfolder it prob-
ably is not supported. Not supported  interfaces  are:  loopback,
non-ethernet and batman's own interfaces.

Note:  After the module was loaded it will continuously watch for
new interfaces to verify the compatibility. There is no  need  to
reload the module if you plug your USB wifi adapter into your ma-
chine after batman advanced was initially loaded.

To activate a  given  interface  simply  write  "bat0"  into  its
"mesh_iface" file inside the batman_adv14 subfolder:

# echo bat0 > /sys/class/net/eth0/batman_adv14/mesh_iface

Repeat  this step for all interfaces you wish to add.  Now batman
starts using/broadcasting on this/these interface(s).

By reading the "iface_status" file you can check its status:

# cat /sys/class/net/eth0/batman_adv14/iface_status
# active

To deactivate an interface you have  to  write  "none"  into  its
"mesh_iface" file:

# echo none > /sys/class/net/eth0/batman_adv14/mesh_iface


All  mesh  wide  settings  can be found in batman's own interface
folder:

# ls /sys/class/net/bat0/mesh/
# aggregated_ogms        gw_bandwidth           log_level
# ap_isolation           gw_mode                orig_interval
# bonding                gw_sel_class           routing_algo
# bridge_loop_avoidance  hop_penalty            vis_mode
# fragmentation


There is a special folder for debugging information:

# ls /sys/kernel/debug/batman_adv14/bat0/
# bla_backbone_table  log                 transtable_global
# bla_claim_table     originators         transtable_local
# gateways            socket              vis_data

Some of the files contain all sort of status information  regard-
ing  the  mesh  network.  For  example, you can view the table of
originators (mesh participants) with:

# cat /sys/kernel/debug/batman_adv14/bat0/originators

Other files allow to change batman's behaviour to better fit your
requirements.  For instance, you can check the current originator
interval (value in milliseconds which determines how often batman
sends its broadcast packets):

# cat /sys/class/net/bat0/mesh/orig_interval
# 1000

and also change its value:

# echo 3000 > /sys/class/net/bat0/mesh/orig_interval

In very mobile scenarios, you might want to adjust the originator
interval to a lower value. This will make the mesh  more  respon-
sive to topology changes, but will also increase the overhead.


USAGE
-----

To  make use of your newly created mesh, batman advanced provides
a new interface "bat0" which you should use from this  point  on.
All  interfaces  added  to  batman  advanced are not relevant any
longer because batman handles them for you. Basically, one "hands
over" the data by using the batman interface and batman will make
sure it reaches its destination.

The "bat0" interface can be used like any  other  regular  inter-
face.  It needs an IP address which can be either statically con-
figured or dynamically (by using DHCP or similar services):

# NodeA: ifconfig bat0 192.168.0.1
# NodeB: ifconfig bat0 192.168.0.2
# NodeB: ping 192.168.0.1

Note:  In  order to avoid problems remove all IP addresses previ-
ously assigned to interfaces now used by batman advanced, e.g.

# ifconfig eth0 0.0.0.0


VISUALIZATION
-------------

If you want topology visualization, at least one mesh  node  must
be configured as VIS-server:

# echo "server" > /sys/class/net/bat0/mesh/vis_mode

Each  node  is  either configured as "server" or as "client" (de-
fault: "client").  Clients send their topology data to the server
next to them, and server synchronize with other servers. If there
is no server configured (default) within the  mesh,  no  topology
information   will  be  transmitted.  With  these  "synchronizing
servers", there can be 1 or more vis servers sharing the same (or
at least very similar) data.

When  configured  as  server,  you can get a topology snapshot of
your mesh:

# cat /sys/kernel/debug/batman_adv14/bat0/vis_data

This raw output is intended to be easily parsable and convertable
with  other tools. Have a look at the batctl README if you want a
vis output in dot or json format for instance and how those  out-
puts could then be visualised in an image.

The raw format consists of comma separated values per entry where
each entry is giving information about a  certain  source  inter-
face.  Each  entry can/has to have the following values:
-> "mac" - mac address of an originator's source interface
           (each line begins with it)
-> "TQ mac  value"  -  src mac's link quality towards mac address
                       of a neighbor originator's interface which
                       is being used for routing
-> "TT mac" - TT announced by source mac
-> "PRIMARY" - this  is a primary interface
-> "SEC mac" - secondary mac address of source
               (requires preceding PRIMARY)

The TQ value has a range from 4 to 255 with 255 being  the  best.
The TT entries are showing which hosts are connected to the mesh
via bat0 or being bridged into the mesh network.  The PRIMARY/SEC
values are only applied on primary interfaces


LOGGING/DEBUGGING
-----------------

All error messages, warnings and information messages are sent to
the kernel log. Depending on your operating  system  distribution
this  can  be read in one of a number of ways. Try using the com-
mands: dmesg, logread, or looking in the files  /var/log/kern.log
or  /var/log/syslog.  All  batman-adv14  messages are prefixed with
"batman-adv14:" So to see just these messages try

# dmesg | grep batman-adv14

When investigating problems with your mesh network  it  is  some-
times  necessary  to see more detail debug messages. This must be
enabled when compiling the batman-adv14 module. When building  bat-
man-adv  as  part of kernel, use "make menuconfig" and enable the
option "B.A.T.M.A.N. debugging".

Those additional  debug messages can be accessed  using a special
file in debugfs

# cat /sys/kernel/debug/batman_adv14/bat0/log

The additional debug output is by default disabled. It can be en-
abled  during run time. Following log_levels are defined:

0 - All  debug  output  disabled
1 - Enable messages related to routing / flooding / broadcasting
2 - Enable messages related to route added / changed / deleted
4 - Enable messages related to translation table operations
8 - Enable messages related to bridge loop avoidance
16 - Enable messaged related to DAT, ARP snooping and parsing
31 - Enable all messages

The debug output can be changed at runtime  using  the  file
/sys/class/net/bat0/mesh/log_level. e.g.

# echo 6 > /sys/class/net/bat0/mesh/log_level

will enable debug messages for when routes change.

Counters for different types of packets entering and leaving the
batman-adv14 module are available through ethtool:

# ethtool --statistics bat0


BATCTL
------

As batman advanced operates on layer 2 all hosts participating in
the  virtual switch are completely transparent for all  protocols
above layer 2. Therefore the common diagnosis tools do  not  work
as  expected.  To  overcome these problems batctl was created. At
the  moment the  batctl contains ping,  traceroute,  tcpdump  and
interfaces to the kernel module settings.

For more information, please see the manpage (man batctl).

batctl is available on http://www.open-mesh.org/


CONTACT
-------

batman-adv14-legacy  is  maintained  by the developers of the Gluon
firmware framework (see https://github.com/freifunk-gluon/gluon).

You  can find us on IRC: (channel #gluon on irc.hackint.org, look
for neoraider)

Bug reports can also be sent via mail to:

Matthias Schiffer <mschiffer@universe-factory.net>


CHANGELOG
batman-adv14 2013.4.0

 * support latest kernels (2.6.29 - 3.12)
 * bugs squashed:
   - fix potential kernel paging errors for unicast transmissions
   - fix network coding initialization for multiple soft interfaces
   - fix BLA VLAN handling regression
   - improve backward compatibility code

 -- Sun, 13 Oct 2013 18:50:11 +0200

batman-adv14 2013.3.0:

 * support latest kernels (2.6.29 - 3.11)
 * send each broadcast only once on non-wireless interfaces
 * change VID handling to prepare enhanced VLAN features
 * bugs squashed:
   - forward late OGMs from best next hop
   - avoid double free of bat_counters
   - fix rcu barrier miss
   - use the proper header len when checking the TTVN
   - make DAT drop ARP requests targeting local clients
 * style improvements, code moving and refactoring

 -- Sat, 20 Jul 2013 22:06:46 +0200

batman-adv14 2013.2.0:

 * support latest kernels (2.6.29 - 3.10)
 * add network coding feature
 * add rtnl interface configuration support
 * fix rtnl and sysfs locking issue
 * avoid duplicate interface enslaving
 * bugs squashed:
   - verify tt len to not exceed packet len
   - fix identification of own mac addresses
 * style improvements (change seq_printf to seq_puts)

 -- Sat, 20 Apr 2013 16:42:51 +0200

batman-adv14 2013.1.0:

 * support latest kernels (2.6.29 - 3.9)
 * add lots of kerneldoc in types.h
 * clean up (kerneldoc alignment, group tt definitions, renaming, ...)
 * TT improvements:
   - add CRC to debugging tables
   - ignore multicast addresses
   - reduce local TT timeout from 1 hour to 10 minutes
 * Initialize lockdep class keys for hashes
 * unbloat bat_priv if debug is not enabled
 * bugs squashed
   - fix possible sysfs/rtnl deadlock when deregistering
   - fix some DAT bugs (skb leak, invalid MAC addresses, NULL pointer dereference)

 -- Sat, 02 Mar 2013 16:02:00 +0100

batman-adv14 2013.0.0:

* support latest kernels (2.6.29 - 3.8)
* cleanup of coding style and add kerneldoc
* add new distributed ARP table feature to cache ARP entries
* remove __packed attribute whenever possible to allow better compiler
  optimizations
* use kernel-registered BATMAN Ethernet type
* block ECTP traffic to remove integration problems when using BLA
* allow roaming multiple times in TT
* bugs squashed
  - fix lockdeps for interfaces
  - disallow batman-over-batman configurations
  - various BLA fixes (wait at startup, hashing, duplist, ...)
  - fix TT roaming cases
  - fix TT packet rerouting
  - fix TT flags handling
  - fix speedy join/BLA interaction bug
  - fix random jitter calculation

 -- Sat, 12 Jan 2013 12:00:16 +0100

batman-adv14 2012.4.0:

* support latest kernels (2.6.29 - 3.7)
* cleanup of coding style
* integration of remaining packet counters in ethtool statistic
* speedy join of new non-mesh clients based on broadcast packets
* added bridge loop avoidance backbone gateway debugfs table
* workaround for kernel bug when running on top of vlan device
* bugs squashed
  - announcement of own MAC address after modification
  - wrong counting of received OGMs on x86
  - route flapping in multiple interfaces setup
  - receiving of translation table requests for foreign nodes
  - invalid memory access after failure during the interface appending
  - wrong calculation of packet crc which lead to dropping of broadcast
    packets when bridge loop avoidance is enabled

 -- Mon, 29 Oct 2012 16:47:05 +0100

batman-adv14 2012.3.0:

* support latest kernels (2.6.29 - 3.6)
* added namespace-like prefix for all batman-adv14 symbols
* integrated extended statistic support using ethtool
* important bugs fixed
  - correct endianness for translation table crc
  - avoid race condition in translation table replacements
  - ensure gateway gets selected
  - allow broadcasts with enabled AP isolation
  - fix vis output for multiple interface configurations
  - fix race condition during adds in hashes
  - fix dropped DHCP packets with enabled bridge loop avoidance and gateway
    support
  - don't leak information through uninitialized packets fields

 -- Sun, 19 Aug 2012 21:28:23 +0200

batman-adv14 2012.2.0:

* support latest kernels (2.6.29 - 3.5)
* cleanup of coding style (use of named constants instead of values, refactored
  code to reduce readability, replace bitarrays with kernel functionality, ...)
* tweaking hop penalty to reduce route flapping
* enhanced the framework to support multiple routing algorithms
* reimplemented the bridge loop avoidance with support for multiple active
  backbone gateways
* some bugs fixed (translation table flag handling, correct unicast rerouting,
  endianness fixed for translation table crc, avoid of routing loops by being
  strict on forwarded ogms, ...)

 -- Tue, 12 Jun 2012 18:21:54 +0200

batman-adv14 2012.1.0:

* support latest kernels (2.6.29 - 3.4)
* makefile rewrite (install target, cleanups, new selection system for features)
* cleanup of coding style (explicit marking of common headers, common unit for
  time specific defines, ...)
* Added framework to switch between different routing algorithms
* some bugs fixes (distinguish between wrap-around ttvn and uninitialized ttvn
  value, ...)

 -- Fri, 30 Mar 2012 19:31:12 +0300

batman-adv14 2012.0.0:

* support latest kernels (2.6.29 - 3.3)
* Fix bat_socket_read memory corruption (CVE-2011-4604)
* Cleanup of gateway handling code
* many bugs (hang when softif creation fails, memory leaks when hashes
  table cannot be filled, wrong filter for missed ogms, many smaller
  translation table problems, ...) fixed

 -- Sun, 05 Feb 2012 15:06:08 +0100

batman-adv14 2011.4.0:

* support latest kernels (2.6.29 - 3.2)
* starting of code refactoring to support multiple protocol versions
* added support for AP-isolation to prevent non-mesh WIFI clients to talk to
  each other over the mesh
* some bugs (memory leaks in the translation table, wrong initialization of
  ethernet addresses of translation table entries, ...) squashed

 -- Sun, 13 Nov 2011 20:40:05 +0100


main.h
/* Copyright (C) 2007-2013 B.A.T.M.A.N. contributors:
 *
 * Marek Lindner, Simon Wunderlich
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA
 */

#ifndef _NET_BATMAN_ADV14_MAIN_H_
#define _NET_BATMAN_ADV14_MAIN_H_

#define BATADV_DRIVER_AUTHOR "Marek Lindner <lindner_marek@yahoo.de>, " \
			     "Simon Wunderlich <siwu@hrz.tu-chemnitz.de>"
#define BATADV_DRIVER_DESC   "B.A.T.M.A.N. advanced"
#define BATADV_DRIVER_DEVICE "batman-adv14"

#ifndef BATADV_SOURCE_VERSION
#define BATADV_SOURCE_VERSION "2013.4.0"
#endif

/* B.A.T.M.A.N. parameters */

#define BATADV_TQ_MAX_VALUE 255
#define BATADV_JITTER 20

/* Time To Live of broadcast messages */
#define BATADV_TTL 50

/* purge originators after time in seconds if no valid packet comes in
 * -> TODO: check influence on BATADV_TQ_LOCAL_WINDOW_SIZE
 */
#define BATADV_PURGE_TIMEOUT 200000 /* 200 seconds */
#define BATADV_TT_LOCAL_TIMEOUT 600000 /* in milliseconds */
#define BATADV_TT_CLIENT_ROAM_TIMEOUT 600000 /* in milliseconds */
#define BATADV_TT_CLIENT_TEMP_TIMEOUT 600000 /* in milliseconds */
#define BATADV_TT_WORK_PERIOD 5000 /* 5 seconds */
#define BATADV_ORIG_WORK_PERIOD 1000 /* 1 second */
#define BATADV_DAT_ENTRY_TIMEOUT (5*60000) /* 5 mins in milliseconds */
/* sliding packet range of received originator messages in sequence numbers
 * (should be a multiple of our word size)
 */
#define BATADV_TQ_LOCAL_WINDOW_SIZE 64
/* milliseconds we have to keep pending tt_req */
#define BATADV_TT_REQUEST_TIMEOUT 3000

#define BATADV_TQ_GLOBAL_WINDOW_SIZE 5
#define BATADV_TQ_LOCAL_BIDRECT_SEND_MINIMUM 1
#define BATADV_TQ_LOCAL_BIDRECT_RECV_MINIMUM 1
#define BATADV_TQ_TOTAL_BIDRECT_LIMIT 1

/* number of OGMs sent with the last tt diff */
#define BATADV_TT_OGM_APPEND_MAX 3

/* Time in which a client can roam at most ROAMING_MAX_COUNT times in
 * milliseconds
 */
#define BATADV_ROAMING_MAX_TIME 20000
#define BATADV_ROAMING_MAX_COUNT 5

#define BATADV_NO_FLAGS 0

#define BATADV_NULL_IFINDEX 0 /* dummy ifindex used to avoid iface checks */

#define BATADV_NUM_WORDS BITS_TO_LONGS(BATADV_TQ_LOCAL_WINDOW_SIZE)

#define BATADV_LOG_BUF_LEN 8192	  /* has to be a power of 2 */

/* number of packets to send for broadcasts on different interface types */
#define BATADV_NUM_BCASTS_DEFAULT 1
#define BATADV_NUM_BCASTS_WIRELESS 3
#define BATADV_NUM_BCASTS_MAX 3

/* msecs after which an ARP_REQUEST is sent in broadcast as fallback */
#define ARP_REQ_DELAY 250
/* numbers of originator to contact for any PUT/GET DHT operation */
#define BATADV_DAT_CANDIDATES_NUM 3

/* how much worse secondary interfaces may be to be considered as bonding
 * candidates
 */
#define BATADV_BONDING_TQ_THRESHOLD	50

/* should not be bigger than 512 bytes or change the size of
 * forw_packet->direct_link_flags
 */
#define BATADV_MAX_AGGREGATION_BYTES 512
#define BATADV_MAX_AGGREGATION_MS 100

#define BATADV_BLA_PERIOD_LENGTH	10000	/* 10 seconds */
#define BATADV_BLA_BACKBONE_TIMEOUT	(BATADV_BLA_PERIOD_LENGTH * 3)
#define BATADV_BLA_CLAIM_TIMEOUT	(BATADV_BLA_PERIOD_LENGTH * 10)
#define BATADV_BLA_WAIT_PERIODS		3

#define BATADV_DUPLIST_SIZE		16
#define BATADV_DUPLIST_TIMEOUT		500	/* 500 ms */
/* don't reset again within 30 seconds */
#define BATADV_RESET_PROTECTION_MS 30000
#define BATADV_EXPECTED_SEQNO_RANGE	65536

#define BATADV_NC_NODE_TIMEOUT 10000 /* Milliseconds */

enum batadv14_mesh_state {
	BATADV_MESH_INACTIVE,
	BATADV_MESH_ACTIVE,
	BATADV_MESH_DEACTIVATING,
};

#define BATADV_BCAST_QUEUE_LEN		256
#define BATADV_BATMAN_QUEUE_LEN	256

enum batadv14_uev_action {
	BATADV_UEV_ADD = 0,
	BATADV_UEV_DEL,
	BATADV_UEV_CHANGE,
};

enum batadv14_uev_type {
	BATADV_UEV_GW = 0,
};

#define BATADV_GW_THRESHOLD	50

#define BATADV_DAT_CANDIDATE_NOT_FOUND	0
#define BATADV_DAT_CANDIDATE_ORIG	1

/* Debug Messages */
#ifdef pr_fmt
#undef pr_fmt
#endif
/* Append 'batman-adv14: ' before kernel messages */
#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt

/* Kernel headers */

#include <linux/mutex.h>	/* mutex */
#include <linux/module.h>	/* needed by all modules */
#include <linux/netdevice.h>	/* netdevice */
#include <linux/etherdevice.h>  /* ethernet address classification */
#include <linux/if_ether.h>	/* ethernet header */
#include <linux/poll.h>		/* poll_table */
#include <linux/kthread.h>	/* kernel threads */
#include <linux/pkt_sched.h>	/* schedule types */
#include <linux/workqueue.h>	/* workqueue */
#include <linux/percpu.h>
#include <linux/slab.h>
#include <net/sock.h>		/* struct sock */
#include <net/rtnetlink.h>
#include <linux/jiffies.h>
#include <linux/seq_file.h>
#include "compat.h"

#include "types.h"

/**
 * batadv14_vlan_flags - flags for the four MSB of any vlan ID field
 * @BATADV_VLAN_HAS_TAG: whether the field contains a valid vlan tag or not
 */
enum batadv14_vlan_flags {
	BATADV_VLAN_HAS_TAG	= BIT(15),
};

#define BATADV_PRINT_VID(vid) (vid & BATADV_VLAN_HAS_TAG ? \
			       (int)(vid & VLAN_VID_MASK) : -1)

extern char batadv14_routing_algo[];
extern struct list_head batadv14_hardif_list;

extern unsigned char batadv14_broadcast_addr[];
extern struct workqueue_struct *batadv14_event_workqueue;

int batadv14_mesh_init(struct net_device *soft_iface);
void batadv14_mesh_free(struct net_device *soft_iface);
int batadv14_is_my_mac(struct batadv14_priv *bat_priv, const uint8_t *addr);
struct batadv14_hard_iface *
batadv14_seq_print_text_primary_if_get(struct seq_file *seq);
int batadv14_batman_skb_recv(struct sk_buff *skb, struct net_device *dev,
			   struct packet_type *ptype,
			   struct net_device *orig_dev);
int
batadv14_recv_handler_register(uint8_t packet_type,
			     int (*recv_handler)(struct sk_buff *,
						 struct batadv14_hard_iface *));
void batadv14_recv_handler_unregister(uint8_t packet_type);
int batadv14_algo_register(struct batadv14_algo_ops *bat_algo_ops);
int batadv14_algo_select(struct batadv14_priv *bat_priv, char *name);
int batadv14_algo_seq_print_text(struct seq_file *seq, void *offset);
__be32 batadv14_skb_crc32(struct sk_buff *skb, u8 *payload_ptr);

/**
 * enum batadv14_dbg_level - available log levels
 * @BATADV_DBG_BATMAN: OGM and TQ computations related messages
 * @BATADV_DBG_ROUTES: route added / changed / deleted
 * @BATADV_DBG_TT: translation table messages
 * @BATADV_DBG_BLA: bridge loop avoidance messages
 * @BATADV_DBG_DAT: ARP snooping and DAT related messages
 * @BATADV_DBG_NC: network coding related messages
 * @BATADV_DBG_ALL: the union of all the above log levels
 */
enum batadv14_dbg_level {
	BATADV_DBG_BATMAN = BIT(0),
	BATADV_DBG_ROUTES = BIT(1),
	BATADV_DBG_TT	  = BIT(2),
	BATADV_DBG_BLA    = BIT(3),
	BATADV_DBG_DAT    = BIT(4),
	BATADV_DBG_NC	  = BIT(5),
	BATADV_DBG_ALL    = 63,
};

#ifdef CONFIG_BATMAN_ADV14_DEBUG
int batadv14_debug_log(struct batadv14_priv *bat_priv, const char *fmt, ...)
__printf(2, 3);

/* possibly ratelimited debug output */
#define _batadv14_dbg(type, bat_priv, ratelimited, fmt, arg...)			\
	do {							\
		if (atomic_read(&bat_priv->log_level) & type && \
		    (!ratelimited || net_ratelimit()))	\
			batadv14_debug_log(bat_priv, fmt, ## arg);\
	}							\
	while (0)
#else /* !CONFIG_BATMAN_ADV14_DEBUG */
__printf(4, 5)
static inline void _batadv14_dbg(int type __always_unused,
			      struct batadv14_priv *bat_priv __always_unused,
			      int ratelimited __always_unused,
			      const char *fmt __always_unused, ...)
{
}
#endif

#define batadv14_dbg(type, bat_priv, arg...) \
	_batadv14_dbg(type, bat_priv, 0, ## arg)
#define batadv14_dbg_ratelimited(type, bat_priv, arg...) \
	_batadv14_dbg(type, bat_priv, 1, ## arg)

#define batadv14_info(net_dev, fmt, arg...)				\
	do {								\
		struct net_device *_netdev = (net_dev);                 \
		struct batadv14_priv *_batpriv = netdev_priv(_netdev);    \
		batadv14_dbg(BATADV_DBG_ALL, _batpriv, fmt, ## arg);	\
		pr_info("%s: " fmt, _netdev->name, ## arg);		\
	} while (0)
#define batadv14_err(net_dev, fmt, arg...)				\
	do {								\
		struct net_device *_netdev = (net_dev);                 \
		struct batadv14_priv *_batpriv = netdev_priv(_netdev);    \
		batadv14_dbg(BATADV_DBG_ALL, _batpriv, fmt, ## arg);	\
		pr_err("%s: " fmt, _netdev->name, ## arg);		\
	} while (0)

/* returns 1 if they are the same ethernet addr
 *
 * note: can't use compare_ether_addr() as it requires aligned memory
 */
static inline int batadv14_compare_eth(const void *data1, const void *data2)
{
	return (memcmp(data1, data2, ETH_ALEN) == 0 ? 1 : 0);
}

/**
 * has_timed_out - compares current time (jiffies) and timestamp + timeout
 * @timestamp:		base value to compare with (in jiffies)
 * @timeout:		added to base value before comparing (in milliseconds)
 *
 * Returns true if current time is after timestamp + timeout
 */
static inline bool batadv14_has_timed_out(unsigned long timestamp,
					unsigned int timeout)
{
	return time_is_before_jiffies(timestamp + msecs_to_jiffies(timeout));
}

#define batadv14_atomic_dec_not_zero(v)	atomic_add_unless((v), -1, 0)

/* Returns the smallest signed integer in two's complement with the sizeof x */
#define batadv14_smallest_signed_int(x) (1u << (7u + 8u * (sizeof(x) - 1u)))

/* Checks if a sequence number x is a predecessor/successor of y.
 * they handle overflows/underflows and can correctly check for a
 * predecessor/successor unless the variable sequence number has grown by
 * more then 2**(bitwidth(x)-1)-1.
 * This means that for a uint8_t with the maximum value 255, it would think:
 *  - when adding nothing - it is neither a predecessor nor a successor
 *  - before adding more than 127 to the starting value - it is a predecessor,
 *  - when adding 128 - it is neither a predecessor nor a successor,
 *  - after adding more than 127 to the starting value - it is a successor
 */
#define batadv14_seq_before(x, y) ({typeof(x) _d1 = (x); \
				 typeof(y) _d2 = (y); \
				 typeof(x) _dummy = (_d1 - _d2); \
				 (void) (&_d1 == &_d2); \
				 _dummy > batadv14_smallest_signed_int(_dummy); })
#define batadv14_seq_after(x, y) batadv14_seq_before(y, x)

/* Stop preemption on local cpu while incrementing the counter */
static inline void batadv14_add_counter(struct batadv14_priv *bat_priv, size_t idx,
				      size_t count)
{
	this_cpu_add(bat_priv->bat_counters[idx], count);
}

#define batadv14_inc_counter(b, i) batadv14_add_counter(b, i, 1)

/* Sum and return the cpu-local counters for index 'idx' */
static inline uint64_t batadv14_sum_counter(struct batadv14_priv *bat_priv,
					  size_t idx)
{
	uint64_t *counters, sum = 0;
	int cpu;

	for_each_possible_cpu(cpu) {
		counters = per_cpu_ptr(bat_priv->bat_counters, cpu);
		sum += counters[idx];
	}

	return sum;
}

/* Define a macro to reach the control buffer of the skb. The members of the
 * control buffer are defined in struct batadv14_skb_cb in types.h.
 * The macro is inspired by the similar macro TCP_SKB_CB() in tcp.h.
 */
#define BATADV_SKB_CB(__skb)       ((struct batadv14_skb_cb *)&((__skb)->cb[0]))

#endif /* _NET_BATMAN_ADV14_MAIN_H_ */
